wrappers/services.go

```diff
@@ -244,139 +242,97 @@ func StartServices(apps []*snap.AppInfo, disabledSvcs *DisabledServices, opts *S
 	}
 
 	systemSysd := systemd.New(systemd.SystemMode, inter)
-	userSysd := systemd.New(systemd.GlobalUserMode, inter)
+	userGlobalSysd := systemd.New(systemd.GlobalUserMode, inter)
 	cli, err := newUserServiceClientNames(opts.Users, inter)
 	if err != nil {
 		return err
 	}
 
-	var toEnableSystem []string
-	var toEnableUser []string
-	systemServices := make([]string, 0, len(apps))
-	userServices := make([]string, 0, len(apps))
-	servicesStarted := false
+	// When starting and enabling services, act on the activated units instead of
+	// the services activated by those units. And since 'static' service units does
+	// not need to be enabled, we can save that.
+	const includeActivatedServices = false
 
+	filteredApps := filterServicesForStart(apps, disabledSvcs, opts.Scope)
+	systemServices, userServices := categorizeServices(filteredApps, includeActivatedServices)
+	var undoStart bool
 	defer func() {
 		if err == nil {
 			return
 		}
-		// apps could have been sorted according to their startup
-		// ordering, stop them in reverse order
-		if servicesStarted {
-			for i := len(apps) - 1; i >= 0; i-- {
-				app := apps[i]
+
+		// Undo logic for user services is handled by user session agent,
+		// we only handle undo logic for system services in this function
+		if undoStart && len(systemServices) > 0 {
+			// filteredApps could have been sorted according to their startup
+			// ordering, stop them in reverse order
+			for i, j := 0, len(filteredApps)-1; i < j; i, j = i+1, j-1 {
+				filteredApps[i], filteredApps[j] = filteredApps[j], filteredApps[i]
+			}
+
+			// Stop them one-by-one to maintain order, the issue is if we just send all of them down
+			// to systemd, it will spawn a process for each stop anyway, just simultaneously.
+			for _, app := range filteredApps {
```

Comment:
could we simply process systemServices but do it in reverse order?

Reply:
We can't, it would not give the correct behaviour -

systemServices will look like this [svc1.sock1, svc1, svc2, svc3]

If we reverse this, then we will end up like this [svc3, svc2, svc1, svc1.sock1] - the issue here being that we stop svc1 before we stop svc1.sock1 - this could potentionally end up starting svc1 back up before we get svc1.sock1 stopped. So we must reverse the stopping order only by app, not by service

Reply:
Okay so initially I thought it included the primary units (the list in systemServices), but I was confusing myself - let me just double check whether this can be done instead

Reply:
Right, the problem is that `systemServices` does not contain the primary units, only the activator units (sockets, timers etc) - when stopping we want to ensure we stop the primary unit as well as it may have been started before we get to the undo